test/python/circuit/test_parameters.py

```diff
@@ -346,6 +346,23 @@ def test_multiple_named_parameters(self):
         self.assertEqual(theta.name, "θ")
         self.assertEqual(qc.parameters, {theta, x})
 
+    @named_data(
+        ["int", 2, int],
+        ["float", 2.1, float],
+        ["float16", numpy.float16(2.1), float],
+        ["float32", numpy.float32(2.1), float],
+        ["float64", numpy.float64(2.1), float],
+    )
+    def test_circuit_assignment_to_numeric(self, value, type_):
+        """Test binding a numeric value to a circuit instruction"""
+        x = Parameter("x")
+        qc = QuantumCircuit(1)
+        qc.append(Instruction("inst", 1, 0, [x]), (0,))
+        qc.assign_parameters({x: value}, inplace=True)
+        bound = qc.data[0].operation.params[0]
+        self.assertIsInstance(bound, type_)
+        self.assertAlmostEqual(bound, value, delta=1e-4)
```

Comment:
Interestingly, this passed for me with `assertEqual` as well and I could not figure out why. I stepped in with a debugger and could see that `bound` and `value` had different precision for the float16 case.

Reply:
Yeah, the closest double-precision float to `float16(2.1)` is different to the closest double-precision float to `real(2.1)`, so the rounding goes differently.  Since this test is just testing types, it'd probably be best to test something with an exact representation in half-precision, such as 2.5, then we can use `assertEqual` to assert that there's no dodgy conversions happening.